service/storage_service.cc

```diff
@@ -1191,6 +1197,69 @@ class topology_coordinator {
         co_return std::tuple{gen_uuid, std::move(guard), std::move(updates.back())};
     }
 
+    // If there are some unpublished CDC generations, publishes the one with the oldest timestamp
+    // to user-facing description tables.
+    future<> publish_oldest_cdc_generation(group0_guard guard) {
+        const auto& unpublished_gens = _topo_sm._topology.unpublished_cdc_generations;
+        if (unpublished_gens.empty()) {
+            co_return;
+        }
+
+        // The generation under index 0 is the oldest because unpublished_cdc_generations are sorted by timestamp.
+        auto gen_id = unpublished_gens[0];
+
+        auto gen_data = co_await _sys_ks.read_cdc_generation(gen_id.id);
+
+        co_await _sys_dist_ks.local().create_cdc_desc(
+                gen_id.ts, gen_data, { get_token_metadata().count_normal_token_owners() });
+
+        std::vector<cdc::generation_id_v2> new_unpublished_gens(unpublished_gens.begin() + 1, unpublished_gens.end());
+        topology_mutation_builder builder(guard.write_timestamp());
+        builder.set_unpublisheded_cdc_generations(std::move(new_unpublished_gens));
+
+        auto str = ::format("published CDC generation, ID: {}", gen_id);
+        co_await update_topology_state(std::move(guard), {builder.build()}, std::move(str));
+    }
+
+    // The background fiber of the topology coordinator that continually publishes committed yet unpublished
+    // CDC generations. Every generation is published in a separate group 0 operation.
+    future<> cdc_generation_publisher_fiber() {
+        slogger.trace("raft topology: start CDC generation publisher fiber");
+
+        while (!_as.abort_requested()) {
+            bool sleep = false;
+            try {
+                auto guard = co_await start_operation();
+
+                co_await publish_oldest_cdc_generation(std::move(guard));
+
+                if (_topo_sm._topology.unpublished_cdc_generations.empty()) {
+                    // No CDC generations to publish. Wait until one appears or the topology coordinator aborts.
+                    slogger.trace("raft topology: CDC generation publisher fiber has nothing to do. Sleeping.");
+                    co_await _topo_sm.event.when([&] () {
+                        return !_topo_sm._topology.unpublished_cdc_generations.empty() || _as.abort_requested();
+                    });
+                    slogger.trace("raft topology: CDC generation publisher fiber wakes up");
+                }
+            } catch (raft::request_aborted&) {
+                slogger.debug("raft topology: CDC generation publisher fiber aborted");
+            } catch (term_changed_error&) {
```

Comment:
I do not think we should ever get it here. Happens when we try to retake a node.

Reply:
No, I am taking this back. start_operation() may throw it. _as should be aborted at this point, but we do not know what happens first us checking _as or getting the exception.